File: src/test/java/funding/startreum/domain/users/UserControllerTest.java
package funding.startreum.domain.users;

import funding.startreum.common.util.JwtUtil;
import funding.startreum.domain.users.controller.UserController;
import funding.startreum.domain.users.dto.UserResponse;
import funding.startreum.domain.users.repository.RefreshTokenRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@ExtendWith(MockitoExtension.class)  // Mockito 확장 사용
public class UserControllerTest {

    private MockMvc mockMvc;

    @Mock
    private UserService userService;

    @Mock
    private JwtUtil jwtUtil;

    @Mock
    private RefreshTokenRepository refreshTokenRepository;

    @InjectMocks
    private UserController userController;

    @BeforeEach
    void setUp() {
        mockMvc = MockMvcBuilders.standaloneSetup(userController).build();
    }

    @Test
    @DisplayName("ID 중복 확인 API - 중복 아닐 때")
    void checkNameDuplicate_noDuplicate() throws Exception {
        String testName = "testUser";
        when(userService.isNameDuplicate(testName)).thenReturn(false);

        mockMvc.perform(get("/api/users/check-name")
                        .param("name", testName))
                .andExpect(status().isOk())
                .andExpect(content().string("false"));
    }

    @Test
    @DisplayName("회원가입 API - 성공")
    void registerUser_success() throws Exception {
        mockMvc.perform(post("/api/users/registrar")
                        .param("name", "newUser")
                        .param("email", "newuser@example.com")
                        .param("password", "password123")
                        .param("role", "SPONSOR"))
                .andExpect(status().isFound())
                .andExpect(header().string("Location", "/"));
    }

    @Test
    @DisplayName("로그인 API - 성공")
    void loginUser_success() throws Exception {
        // Mock UserResponse 객체 생성
        UserResponse mockUserResponse = new UserResponse("newUser", "newuser@example.com", User.Role.SPONSOR, null, null);

        // userService.authenticateUser()가 호출되면 mockUserResponse 반환
        when(userService.authenticateUser("newUser", "password123")).thenReturn(mockUserResponse);

        // jwtUtil.generateAccessToken()와 generateRefreshToken() 설정
        when(jwtUtil.generateAccessToken(anyString(), anyString(), anyString())).thenReturn("mockAccessToken");
        when(jwtUtil.generateRefreshToken(anyString())).thenReturn("mockRefreshToken");

        // refreshTokenRepository.deleteByUsername()는 void 반환 -> doNothing() 사용
        doNothing().when(refreshTokenRepository).deleteByUsername(anyString());

        // 실제 테스트 실행
        mockMvc.perform(post("/api/users/login")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"name\":\"newUser\",\"password\":\"password123\"}"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.accessToken").value("mockAccessToken"))
                .andExpect(jsonPath("$.refreshToken").value("mockRefreshToken"))
                .andExpect(jsonPath("$.userName").value("newUser"));
    }
}

File: src/test/java/funding/startreum/domain/users/controller/UserControllerTest.kt
package funding.startreum.domain.users.controller

import com.fasterxml.jackson.databind.ObjectMapper
import funding.startreum.common.config.TestSecurityConfig
import funding.startreum.common.util.JwtUtil
import funding.startreum.domain.users.dto.LoginRequest
import funding.startreum.domain.users.dto.SignupRequest
import funding.startreum.domain.users.dto.UserResponse
import funding.startreum.domain.users.entity.User
import funding.startreum.domain.users.repository.RefreshTokenRepository
import funding.startreum.domain.users.service.MyFundingService
import funding.startreum.domain.users.service.MyProjectService
import funding.startreum.domain.users.service.UserService
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mockito.doNothing
import org.mockito.Mockito.`when`
import org.mockito.MockitoAnnotations
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.context.annotation.Import
import org.springframework.http.MediaType
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
import java.time.LocalDateTime

@ExtendWith(SpringExtension::class, MockitoExtension::class)
@WebMvcTest(UserController::class)
@Import(TestSecurityConfig::class)
class UserControllerTest {

    @Autowired
    private lateinit var mockMvc: MockMvc

    @Autowired
    private lateinit var objectMapper: ObjectMapper

    @MockitoBean
    private lateinit var userService: UserService

    @MockitoBean
    private lateinit var jwtUtil: JwtUtil  // ✅ @Mock -> @MockitoBean으로 변경

    @MockitoBean
    private lateinit var myFundingService: MyFundingService

    @MockitoBean
    private lateinit var myProjectService: MyProjectService

    @MockitoBean
    private lateinit var refreshTokenRepository: RefreshTokenRepository

    @BeforeEach
    fun setup() {
        MockitoAnnotations.openMocks(this)
    }



    @Test
    fun `회원가입 성공`() {
        val request = SignupRequest("testUser", "test@example.com", "password123", User.Role.valueOf("SPONSOR"))


        doNothing().`when`(userService).registerUser(request)

        mockMvc.perform(post("/api/users/registrar")
            .param("name", request.name)
            .param("email", request.email)
            .param("password", request.password)
            .param("role", request.role?.name ?: "USER")
        )
            .andExpect(status().isFound)
            .andExpect(header().string("Location", "/"))
    }



    @Test
    fun `로그인 성공`() {
        val request = LoginRequest("testUser", "password123")
        val userResponse = UserResponse(
            name = "testUser",
            email = "test@example.com",
            role = User.Role.SPONSOR, // ✅ 존재하는 값 사용
            createdAt = LocalDateTime.now(),
            updatedAt = LocalDateTime.now()
        )
        val accessToken = "access-token"
        val refreshToken = "refresh-token"

        `when`(userService.authenticateUser(request.name, request.password)).thenReturn(userResponse)
        `when`(jwtUtil.generateAccessToken(userResponse.name, userResponse.email, userResponse.role?.name ?: "USER")).thenReturn(accessToken)
        `when`(jwtUtil.generateRefreshToken(userResponse.name)).thenReturn(refreshToken)

        mockMvc.perform(post("/api/users/login")
            .contentType(MediaType.APPLICATION_JSON)
            .content(objectMapper.writeValueAsString(request))
        )
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.accessToken").value(accessToken))
            .andExpect(jsonPath("$.refreshToken").value(refreshToken))
    }

    @Test
    fun `로그아웃 성공`() {
        mockMvc.perform(post("/api/users/logout"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.status").value("success"))
    }

    @Test
    fun `사용자 프로필 조회 성공`() {
        val user = User("testUser", "test@example.com", "password123", User.Role.SPONSOR, LocalDateTime.now(), LocalDateTime.now())

        `when`(userService.getUserByName("testUser")).thenReturn(user) // ✅ User 반환

        mockMvc.perform(get("/api/users/profile/testUser"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.status").value("success"))
            .andExpect(jsonPath("$.data.name").value(user.name)) // ✅ user.name 사용
    }
}
